### ISSUE TITLE
wxRichTextCtrl may omit some paragraph spacing after the control is resized

### ISSUE THREAD

**Issue by mbc-one** [wxWidgets/wxWidgets#23828]
Describe the bug:
In a wxRichTextCtrl where paragraph spacing is in effect, if the control is resized, the text lines may be rendered with one less spacing pixel line than is specified.

Expected vs observed behaviour:
It is expected that when the control is resized, the spacing between lines of text remains the same (before and after resizing).
In some circumstances, one pixel line is omitted.

Patch or snippet allowing to reproduce the problem:
The patch creates a series of wxRichTextCtrl with various paragraph spacing and resizes some of them, showing visual comparison of unresized and resized controls.

To Reproduce:
1. Use the sample: minimal
2. Do the following in the file:  minimal.cpp
3. At the top, include a header file:   wx/richtext/richtextctrl.h
4. In the function MyFrame::MyFrame() i.e. constructor,
  keep everything as is, then insert at the end of the function (just before the closing brace)
  the following code:
    
```
#pragma comment(lib,"wxmsw32u_richtext.lib")
#pragma comment(lib,"wxbase32u_xml.lib")
#pragma comment(lib,"wxmsw32u_html.lib")

    this->SetBackgroundColour(*wxLIGHT_GREY);

    // wxPanel does not resize it's child windows
    wxPanel *pan = new wxPanel(
        this, wxID_ANY,
        wxPoint(0, 0), wxSize(100, 100),
        wxTAB_TRAVERSAL, wxEmptyString);

    // a throwaway control, off screen, just to get font character size in advance
    wxRichTextCtrl *rt = new wxRichTextCtrl(
        pan, wxID_ANY, wxEmptyString,
        wxPoint(-400, -400), wxSize(100, 100),
        wxRE_READONLY | wxBORDER_NONE | wxWANTS_CHARS);

    int h = rt->GetCharHeight();
    int w = rt->GetCharWidth();

    int rt_w = w * 8;
    int rt_h = h * 14;
    int x = 0;
    int gap = w;
    wxColour bg1 = wxColour(255, 255, 255);
    wxColour bg2 = wxColour(255, 255, 240);
    wxRichTextAttr BasicStyle;

    // for each paragraph spacing ps_px (denominated in pixels)
    for (int ps_px = 0; ps_px < 6; ++ps_px)
    {
        // make 2 identical richtext controls. the second will be RESIZED
        for (int j = 0; j < 2; ++j)
        {
            // label the control pairs with paragraph spacing value
            wxStaticText *st = new wxStaticText(pan, wxID_ANY, wxString::Format("%d", ps_px),
                wxPoint(x, 10), wxDefaultSize,
                wxALIGN_CENTRE_HORIZONTAL);

            st->SetSize(rt_w, st->GetSize().y);

            wxRichTextCtrl *rt = new wxRichTextCtrl(
                pan, wxID_ANY, wxEmptyString,
                wxPoint(x, st->GetRect().GetBottom() + h / 2), wxSize(rt_w, rt_h),
                wxRE_READONLY | wxWANTS_CHARS);

            // visually distinguish the control pairs
            st->SetBackgroundColour((ps_px % 2) ? bg2 : bg1);
            rt->SetBackgroundColour((ps_px % 2) ? bg2 : bg1);

            rt->SetMargins(2, 0);

            int ps_mm = rt->GetBuffer().ConvertPixelsToTenthsMM(wxClientDC(rt), ps_px);

            // use ps_mm to set either ParagraphSpacingBefore or ParagraphSpacingAfter
            // the fault occurs in both cases
            BasicStyle = rt->GetBasicStyle();
            BasicStyle.SetParagraphSpacingBefore(0);
            BasicStyle.SetParagraphSpacingAfter(ps_mm);
            rt->SetBasicStyle(BasicStyle);

            for (int i = 1; i <= 10; ++i)
            {
                rt->WriteText(wxString::Format("line %d\n", i));
            }

            // if second richtext control of a pair:  RESIZE
            if (j)
            {
                rt->SetSize(wxSize(rt_w, rt_h + h * 3));

                // workaround. should not be required - remove comment to activate
                //rt->LayoutContent();
            }

            x += gap + rt_w;
        }
    }

this->SetFocus();

```

The screenshot shows the fault occurring when ParagraphSpacingAfter is 3, 4, or 5 pixels, but not for 0, 1, or 2.

![wx richtext](https://github.com/wxWidgets/wxWidgets/assets/143296853/48b3788e-dea8-4f2b-8688-4851716e65ee)

### Platform and version information

wxWidgets version 3.2.2.1
wxWidgets port wxMSW
OS Windows 8.1
Visual Studio 2013 (MSVC 12)

**Comment by vadz** [comment]
Thanks for reporting this but unfortunately I don't know much about wxRTC so I can't really help with fixing this. If you could please try debugging the problem and submitting a PR fixing it, it would be very welcome. TIA!

**Comment by mbc-one** [comment]
My investigation of this issue reveals the discrepancy in paragraph spacing arises in the context of user-set desktop scaling (here 125%) and is caused by drawing measurements being calculated with GetPPI() functions that produce conflicting results.

MSW wxWidgets 3.2.2.1

Windows 8.1
Control Panel - Display
 Change the size of all items
  Medium - 125%

With magic base dpi = 96 dpi 
 125% = 120 dpi

In RichTextCtrl, layout and drawing functions make sizing calculations based on [device context].GetPPI().

Two types of device context are used:
 wxClientDC
 wxBufferedPaintDC
Their GetPPI()'s produce different results.
``````
wxRichTextCtrl *rtc = new wxRichTextCtrl(...);

wxClientDC(rtc).GetPPI().x         :  120   (good)
wxBufferedPaintDC(rtc).GetPPI().x  :   96   (bad)
``````
The explanation appears to be that wxBufferedPaintDC derives from wxMemoryDC
and wxMemoryDC::GetPPI() always returns standard display dpi 96, not user-set dpi.

``````
msw/dcmemory.cpp (146)

wxSize wxMemoryDCImpl::GetPPI() const
{
    return wxDisplay::GetStdPPI() * GetContentScaleFactor();
}


include/wx/display.h (90)

    // get the default resolution for displays on this platform
    static int GetStdPPIValue()
    {
#ifdef __WXOSX__
        return 72;
#else
        return 96;
#endif
    }

    static wxSize GetStdPPI()
    {
        return wxSize(GetStdPPIValue(), GetStdPPIValue());
    }
``````

Possible solutions

1. Modify wxMemoryDC::GetPPI()

2. Modify wxBufferedPaintDC to have it's own GetPPI()

  1 and 2 may have repercussions beyond wxRichTextCtrl.

3. Confined to wxRichTextCtrl
   Modify the functions that use <device context>.GetPPI().

richtextbuffer.cpp
(633) wxRichTextObject::ConvertTenthsMMToPixels(wxDC& dc, ...)
(663) wxRichTextObject::ConvertPixelsToTenthsMM(wxDC& dc, ...)
(14360) wxTextAttrDimensionConverter::wxTextAttrDimensionConverter(wxDC& dc, ...)

The received dc could be either wxClientDC or wxBufferedPaintDC.
Make a wxClientDC to be sure, or use wxDisplay.
``````
 replace: dc.GetPPI()
 with:    wxClientDC(dc.GetWindow()).GetPPI()
   or:    wxDisplay(dc.GetWindow()).GetPPI()
``````
Either of these will fix the issue reported here.
The richtext sample, compiled with the changed library, runs as expected with small changes in the spacing.

richtextstyles.cpp
This has it's own ConvertTenthsMMToPixels (896).
It is used once (758) with a wxClientDC, so the modification shown above is not required, but maybe should be made anyway.

--------

Also, a note about another problem.
wxRichTextBuffer has  GetDimensionScale(), SetDimensionScale() and m_dimensionScale  which is initialised to 1.0.
It is used once in wxRichTextObject::ConvertTenthsMMToPixels()
richtextbuffer.cpp
(638) scale = GetBuffer()->GetScale() / GetBuffer()->GetDimensionScale();
It is not used in ConvertPixelsToTenthsMM()
(668) scale = GetBuffer()->GetScale();
If m_dimensionScale != 1.0 then conversion from Pixels to TenthsMM will be incorrect.
In other words, lines 638 and 668 should read the same (?).

**Comment by vadz** [comment]
Thanks for looking at this!

`wxMemoryDC` can return correct DPI, but for this it needs to be initialized with a bitmap using the correct scale factor (i.. 1.25 in your case) because otherwise it has no way of knowing which DPI to use -- not being associated with any window/screen. So normally it would be the responsibility of whoever created it, but for `wxBufferedPaintDC` this should be done automatically, of course, as it does have the associated window and should use its scale factor.

IOW I think it might not be a wxRTC-specific problem at all but happen for anything using `wxBufferedPaintDC`, which is much worse. I don't see any trivial solution for this, but we do need to ensure the buffer bitmap uses the correct scale for the window it is used with.

**Comment by MaartenBent** [comment]
The `wxBufferedPaintDC` in
https://github.com/wxWidgets/wxWidgets/blob/3861ad6b6d9b202dcb0f1cb56bcd52a7c375290a/src/richtext/richtextctrl.cpp#L451

is created from both a `wxWindow` and a `wxBitmap`. It uses the scale of the bitmap, which is `1.0`, while the scale/DPI of the window can be different.
Creating the bitmap with the correct scale in
https://github.com/wxWidgets/wxWidgets/blob/3861ad6b6d9b202dcb0f1cb56bcd52a7c375290a/src/richtext/richtextctrl.cpp#L3029
using
```c++
    if (!m_bufferBitmap.IsOk() || m_bufferBitmap.GetLogicalWidth() < sz.x || m_bufferBitmap.GetLogicalHeight() < sz.y) {
        // As per https://github.com/wxWidgets/wxWidgets/issues/14403, prevent very inefficient fix to alpha bits of
        // destination by making the backing bitmap 24-bit. Note that using 24-bit depth breaks painting of
        // scrolled areas on wxWidgets 2.8.
        int depth = -1;
#if defined(__WXMSW__)
        depth = 24;
#endif
        m_bufferBitmap.CreateWithDIPSize(sz, GetDPIScaleFactor(), depth);
    }
```
makes the provided same looks correct (I think).
But just like vadz I don't know much about wxRTC either.

**Comment by vadz** [comment]
Thanks, I didn't even check wxRTC sources and thought it was using the default/shared bitmap. When the bitmap is specified like this, it's definitely the callers responsibility to create it with the correct scale factor, so the above looks correct to me. And, worse, looking at the shared buffer case more carefully, I see that we already handle it by using the scale factor of the target DC (which should be the same as that of the window) for it, so the problem I had originally in mind probably doesn't exist at all, sorry for the confusion...

@mbc-one Could you please check if the fix above works? TIA!

**Comment by mbc-one** [comment]
Thanks @MaartenBent and @vadz .
That fixes it.
A question: 
The original code had separate bitmap creation code for non-MSW
`        m_bufferBitmap = wxBitmap(sz.x, sz.y);
`
Is the new code
`        m_bufferBitmap.CreateWithDIPSize(sz, GetDPIScaleFactor(), depth);
`
suitable for non-MSW ?
Documentation for  wxWindow::GetDPIScaleFactor()  has:

> Notice that this method should not be used for window sizes expressed in pixels, as they are already scaled by this factor by the underlying toolkit under some platforms.

**Comment by MaartenBent** [comment]
The original code was specific to the bitmap depth. I refactored that a bit, but it is still there.

`propgrid` does the same when creating its double buffer, so I think it should work. But I'll test it on other platforms like macOS as well before creating a PR.

**Comment by mbc-one** [comment]
Thanks again for your help, and for the great wxWidgets project.

**Comment by vadz** [comment]
This is not fixed yet, so let's reopen it until the PR is merged.
